keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,388 @@
+  Windows node specific labels.
+- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec
+
+The runtimeclass is a higher level abstraction which gets translated again to nodeSelectors+tolerations. While the 
+nodeSelector with reserved OS label is good enough, it can be viewed as an side-effect of scheduling constraints 
+and need not necessarily express the user intention fully. So, we propose to add a new field to the pod spec called 
+os to identify the OS on which pod intends to run.
+
+```go
+type PodSpec struct {
+  // If specified, indicates the type of OS on which pod intends
+  // to run.  This field is alpha-level and is only honored by servers that 
+  // enable the IdentifyOS feature. This is used to help identifying the
+  // OS authoritatively during the API server admission.
+  // +optional
+  OS *OS
+}
+
+// OS has information on the type of OS. 
+// We're making this a struct for possible future expansion.
+type OS struct {
+  // Name of the OS. The values supported are available at:
+  // https://github.com/opencontainers/runtime-spec/blob/master/config.md#platform-specific-configuration
+  Name string
+}
+
+```
+
+With the above change, all the admission plugins which validate or mutate the pod can 
+identify the pod's OS authoritatively and can act accordingly. As of now, PodSecurityAdmission
+plugin is the only admission plugin can leverage this field on the pod spec. In future, we can
+have a validating admission plugin for Windows pods as Linux and Windows host capabilities are
+different and not all Kubernetes features are supported on Windows host.
+
+
+### User Stories (Optional)
+
+
+#### Story 1
+
+As a Kubernetes cluster administrator, I want appropriate security contexts to be 
+applied to my Windows Pods along with Linux pods
+
+#### Story 2
+
+As a Kubernetes cluster administrator, I want to use my own admission webhook for pods
+based on the OS it intends to run on.
+
+### Notes/Constraints/Caveats (Optional)
+
+Since the OS field is optional in the pod spec, we can expect the users to omit this field
+when submitting pod to API server. In this scenario, we expect admission plugins to treat the pod
+the way it is being treated now.
+
+### Risks and Mitigations
+
+The primary risk is a bug in the implementation of the admission plugins that validate or mutate
+based on the OS field in the pod spec. The best mitigation for that scenario is unit testing 
+when this featuregate is enabled and disabled.
+
+
+## Design Details
+
+### Implications to Kubelet
+
+Apart from the above API change, we intend to make the following changes to Kubelet:
+- Kubelet should reject admitting pods if the pod.Spec.OS.Name doesn't match the node label `kubernetes.io/os`
```

Comment:
When/if linux containers on windows are supported, this will need to be updated, correct?  I think it's still ok to add this since old kubelets won't be able to correctly manage such pods, but I want to call out our future debt.

Perhaps it is better phrased as, "kubelets should reject admitting pod if the kubelet cannot honor the pod.spec.os.name. For instance, if the os.name does not match the host os".

Reply:
Agree, matching the `kubernetes.io/os` node label is an implementation detail. For the KEP we can use the phasing mentioned above.